<a href="https://colab.research.google.com/github/sharonma1218/stats-306/blob/main/lecture02_class.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lecture 2: More on data transformations

<div style="border: 1px double black; padding: 10px; margin: 10px">

**Goals for today's lecture:**
* Continue to learn how to manipulate data, including:
    * Vectors in R: what they are and basic operations
    * Adding new variables
    * Grouping and summarizing data
    
This lecture note corresponds to sections 4.3-4.6 of your book.
</div>

In [2]:
install.packages('tidyverse')
install.packages('nycflights13')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [3]:
library(tidyverse)  # always!
library(nycflights13)  # install if needed

Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.2 ──
✔ ggplot2 3.4.0      ✔ purrr   0.3.5 
✔ tibble  3.1.8      ✔ dplyr   1.0.10
✔ tidyr   1.2.1      ✔ stringr 1.4.1 
✔ readr   2.1.3      ✔ forcats 0.5.2 
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


## Review of last lecture
Last lecture we learned how to alter the rows and columns of a dataframe:
- `filter()` to keep certain rows that satisfy logical conditions.
- `arrange()` to sort rows according to certain column values.
- `distinct()` to keep only rows that are distinct on some combination of columns.
- `select()` to drop/rename/rearrange columns

## What's a data frame?

Our main goal in R is to work with data, and one of the most fundamental objects in R is the *data frame*. Think of a data frame as a container for a bunch of *vectors* of data:

![dataframe](https://garrettgman.github.io/images/tidy-2.png)

## What's a vector?

- In programming speak: a *vector* is a list of values. 
- In statistical speak: a vector of observations (aka data).

Let's create a vector and work with it:

## Poll
How old are you?
<ol style="list-style-type: upper-alpha;">
    <li>19 or younger</li>
    <li>20</li>
    <li>21</li>
    <li>22 or older</li>
    <li>I forget</li>
</ol>
(This question will be graded.)<br />
(↑ This is a joke.)

The function for creating a vector in R is called simply, `c()`.

In [4]:
# Create a vector of ages
ages=c(20,21,22,22,24,25,21,20,19,18)
ages

[1] 20 21 22 22 24 25 21 20 19 18

In [5]:
mean(ages)

[1] 21.2

In [6]:
sd(ages)

[1] 2.149935

In [7]:
log1=c(T,F,T)
log1

[1]  TRUE FALSE  TRUE

In [8]:
log1=c(F,F,F)
any(log1) # NOTE: if any of them are true, then the entire expression becomes true

[1] FALSE

In [11]:
log1=c(F,F,T)
any(log1) # NOTE: even if just one of them is true, then the entire expression turns out to be true

[1] TRUE

In [12]:
log1=c(F,F,T)
all(log1) # NOTE: if all of them are true

[1] FALSE

In [13]:
log1=c(T,T,T)
all(log1) 
# NOTE: in a way, any is like or & all is like and 

[1] TRUE

## Functions that operate on vectors

Many summary functions are available:

* Center: `mean(), median()`
* Spread: `sd(), IQR(), mad()`
* Range: `min(), max(), quantile()`
* Position: `first(), last(), nth()`
* Count: `n(), n_distinct()`
* Logical: `any(), all()`

In [ ]:
# examples of functions we can use on the ages vector

## Column operation #2: `mutate()`

`mutate()` creates new columns in a dataframe that are calculated from the existing columns.

For example, let's define the **gain** of a flight to be the different between the departure delay and the arrival delay:

$$\text{gain} = \text{dep. delay} - \text{arr. delay}$$

So, the gain is positive if the flight made up time in the air, resulting in a less-delayed arrival.

To add a column called `gain` to flights, we called `mutate()` as follows:

In [14]:
?flights

In [17]:
glimpse(flights)

Rows: 336,776
Columns: 19
$ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
$ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ day            <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ dep_time       <int> 517, 533, 542, 544, 554, 554, 555, 557, 557, 558, 558, …
$ sched_dep_time <int> 515, 529, 540, 545, 600, 558, 600, 600, 600, 600, 600, …
$ dep_delay      <dbl> 2, 4, 2, -1, -6, -4, -5, -3, -3, -2, -2, -2, -2, -2, -1…
$ arr_time       <int> 830, 850, 923, 1004, 812, 740, 913, 709, 838, 753, 849,…
$ sched_arr_time <int> 819, 830, 850, 1022, 837, 728, 854, 723, 846, 745, 851,…
$ arr_delay      <dbl> 11, 20, 33, -18, -25, 12, 19, -14, -8, 8, -2, -3, 7, -1…
$ carrier        <chr> "UA", "UA", "AA", "B6", "DL", "UA", "B6", "EV", "B6", "…
$ flight         <int> 1545, 1714, 1141, 725, 461, 1696, 507, 5708, 79, 301, 4…
$ tailnum        <chr> "N14228", "N24211", "N619AA", "N804JB", "N668DN", "N394…
$ origin      

In [21]:
head(flights,5)

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1 2013 1     1   517      515             2         830      819          
2 2013 1     1   533      529             4         850      830          
3 2013 1     1   542      540             2         923      850          
4 2013 1     1   544      545            -1        1004     1022          
5 2013 1     1   554      600            -6         812      837          
  arr_delay carrier flight tailnum origin dest air_time distance hour minute
1  11       UA      1545   N14228  EWR    IAH  227      1400     5    15    
2  20       UA      1714   N24211  LGA    IAH  227      1416     5    29    
3  33       AA      1141   N619AA  JFK    MIA  160      1089     5    40    
4 -18       B6       725   N804JB  JFK    BQN  183      1576     5    45    
5 -25       DL       461   N668DN  LGA    ATL  116       762     6     0    
  time_hour          
1 2013-01-01 05:00:00
2 2013-01-01 05:00:00
3 2013-01-01 05:00:00
4 2013-01-01 05:00:00
5 2013-01-01 06:00:00

In [22]:
print(head(flights,5))

# A tibble: 5 × 19
   year month   day dep_time sched_dep…¹ dep_d…² arr_t…³ sched…⁴ arr_d…⁵ carrier
  <int> <int> <int>    <int>       <int>   <dbl>   <int>   <int>   <dbl> <chr>  
1  2013     1     1      517         515       2     830     819      11 UA     
2  2013     1     1      533         529       4     850     830      20 UA     
3  2013     1     1      542         540       2     923     850      33 AA     
4  2013     1     1      544         545      -1    1004    1022     -18 B6     
5  2013     1     1      554         600      -6     812     837     -25 DL     
# … with 9 more variables: flight <int>, tailnum <chr>, origin <chr>,
#   dest <chr>, air_time <dbl>, distance <dbl>, hour <dbl>, minute <dbl>,
#   time_hour <dttm>, and abbreviated variable names ¹​sched_dep_time,
#   ²​dep_delay, ³​arr_time, ⁴​sched_arr_time, ⁵​arr_delay


In [23]:
head(flights,5)%>%print # NOTE: piping 

# A tibble: 5 × 19
   year month   day dep_time sched_dep…¹ dep_d…² arr_t…³ sched…⁴ arr_d…⁵ carrier
  <int> <int> <int>    <int>       <int>   <dbl>   <int>   <int>   <dbl> <chr>  
1  2013     1     1      517         515       2     830     819      11 UA     
2  2013     1     1      533         529       4     850     830      20 UA     
3  2013     1     1      542         540       2     923     850      33 AA     
4  2013     1     1      544         545      -1    1004    1022     -18 B6     
5  2013     1     1      554         600      -6     812     837     -25 DL     
# … with 9 more variables: flight <int>, tailnum <chr>, origin <chr>,
#   dest <chr>, air_time <dbl>, distance <dbl>, hour <dbl>, minute <dbl>,
#   time_hour <dttm>, and abbreviated variable names ¹​sched_dep_time,
#   ²​dep_delay, ³​arr_time, ⁴​sched_arr_time, ⁵​arr_delay


In [19]:
# add a gain column to flights
mutate(flights,gain=dep_delay-arr_delay)

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1      2013 1     1   517      515             2         830      819          
2      2013 1     1   533      529             4         850      830          
3      2013 1     1   542      540             2         923      850          
4      2013 1     1   544      545            -1        1004     1022          
5      2013 1     1   554      600            -6         812      837          
6      2013 1     1   554      558            -4         740      728          
7      2013 1     1   555      600            -5         913      854          
8      2013 1     1   557      600            -3         709      723          
9      2013 1     1   557      600            -3         838      846          
10     2013 1     1   558      600            -2         753      745          
11     2013 1     1   558      600            -2         849      851          
12     2013 1     1   558      600            -2         853      856          
13     2013 1     1   558      600            -2         924      917          
14     2013 1     1   558      600            -2         923      937          
15     2013 1     1   559      600            -1         941      910          
16     2013 1     1   559      559             0         702      706          
17     2013 1     1   559      600            -1         854      902          
18     2013 1     1   600      600             0         851      858          
19     2013 1     1   600      600             0         837      825          
20     2013 1     1   601      600             1         844      850          
21     2013 1     1   602      610            -8         812      820          
22     2013 1     1   602      605            -3         821      805          
23     2013 1     1   606      610            -4         858      910          
24     2013 1     1   606      610            -4         837      845          
25     2013 1     1   607      607             0         858      915          
26     2013 1     1   608      600             8         807      735          
27     2013 1     1   611      600            11         945      931          
28     2013 1     1   613      610             3         925      921          
29     2013 1     1   615      615             0        1039     1100          
30     2013 1     1   615      615             0         833      842          
⋮      ⋮    ⋮     ⋮   ⋮        ⋮              ⋮         ⋮        ⋮             
336747 2013 9     30  2123     2125            -2       2223     2247          
336748 2013 9     30  2127     2129            -2       2314     2323          
336749 2013 9     30  2128     2130            -2       2328     2359          
336750 2013 9     30  2129     2059            30       2230     2232          
336751 2013 9     30  2131     2140            -9       2225     2255          
336752 2013 9     30  2140     2140             0         10       40          
336753 2013 9     30  2142     2129            13       2250     2239          
336754 2013 9     30  2145     2145             0        115      140          
336755 2013 9     30  2147     2137            10         30       27          
336756 2013 9     30  2149     2156            -7       2245     2308          
336757 2013 9     30  2150     2159            -9       2250     2306          
336758 2013 9     30  2159     1845           194       2344     2030          
336759 2013 9     30  2203     2205            -2       2339     2331          
336760 2013 9     30  2207     2140            27       2257     2250          
336761 2013 9     30  2211     2059            72       2339     2242          
336762 2013 9     30  2231     2245           -14       2335     2356          
336763 2013 9     30  2233     2113            80        112       30          
336764 2013 9     30  2235     2001           154         59     2249          
336765 

The returned data set has a new column called `gain` added to the very end. If you want to bring it to the front, you could use `select()` like we learned last lecture:

In [32]:
# use select to move gain to the first column
select(
  mutate(flights,gain=dep_delay-arr_delay),
  gain,everything())%>%print # NOTE: b/c we want to see these specific columns next to each other

# A tibble: 336,776 × 20
    gain  year month   day dep_time sched_dep_…¹ dep_d…² arr_t…³ sched…⁴ arr_d…⁵
   <dbl> <int> <int> <int>    <int>        <int>   <dbl>   <int>   <int>   <dbl>
 1    -9  2013     1     1      517          515       2     830     819      11
 2   -16  2013     1     1      533          529       4     850     830      20
 3   -31  2013     1     1      542          540       2     923     850      33
 4    17  2013     1     1      544          545      -1    1004    1022     -18
 5    19  2013     1     1      554          600      -6     812     837     -25
 6   -16  2013     1     1      554          558      -4     740     728      12
 7   -24  2013     1     1      555          600      -5     913     854      19
 8    11  2013     1     1      557          600      -3     709     723     -14
 9     5  2013     1     1      557          600      -3     838     846      -8
10   -10  2013     1     1      558          600      -2     753     745       8
# …

In [33]:
glimpse(flights) # NOTE: mutate does not change the orig data set 

Rows: 336,776
Columns: 19
$ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
$ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ day            <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ dep_time       <int> 517, 533, 542, 544, 554, 554, 555, 557, 557, 558, 558, …
$ sched_dep_time <int> 515, 529, 540, 545, 600, 558, 600, 600, 600, 600, 600, …
$ dep_delay      <dbl> 2, 4, 2, -1, -6, -4, -5, -3, -3, -2, -2, -2, -2, -2, -1…
$ arr_time       <int> 830, 850, 923, 1004, 812, 740, 913, 709, 838, 753, 849,…
$ sched_arr_time <int> 819, 830, 850, 1022, 837, 728, 854, 723, 846, 745, 851,…
$ arr_delay      <dbl> 11, 20, 33, -18, -25, 12, 19, -14, -8, 8, -2, -3, 7, -1…
$ carrier        <chr> "UA", "UA", "AA", "B6", "DL", "UA", "B6", "EV", "B6", "…
$ flight         <int> 1545, 1714, 1141, 725, 461, 1696, 507, 5708, 79, 301, 4…
$ tailnum        <chr> "N14228", "N24211", "N619AA", "N804JB", "N668DN", "N394…
$ origin      

In [35]:
# to check whether or not gain is in the orig data set, you can also do the following & see error 
select(flights,gain)

ERROR: ignored

## Quiz
What was the most amount of time gained by any flight?
<ol style="list-style-type: upper-alpha;">
    <li>2 hours</li>
    <li>109 minutes</li>
    <li>37 minutes</li>
    <li>37 seconds</li>
    <li>2 days</li>
</ol>

In [40]:
arrange(
  mutate(flights,gain=dep_delay-arr_delay),
  desc(gain))

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1      2013  6    13  1907     1512           235       2134     1928          
2      2013  2    26  1000      900            60       1513     1540          
3      2013  2    23  1226      900           206       1746     1540          
4      2013  5    13  1917     1900            17       2149     2251          
5      2013  2    27   924      900            24       1448     1540          
6      2013  7    14  1917     1829            48       2109     2135          
7      2013  7    17  2004     1930            34       2224     2304          
8      2013 12    27  1719     1648            31       1956     2038          
9      2013  5     2  1947     1949            -2       2209     2324          
10     2013 11    13  2024     2015             9       2251     2354          
11     2013  5     7  1715     1729           -14       1944     2110          
12     2013  5    12  2019     2000            19       2241     2334          
13     2013  5     2  1949     1910            39       2208     2240          
14     2013  5     6  1826     1830            -4       2045     2200          
15     2013  2    26  1335     1335             0       1819     1929          
16     2013  3     1  1528      920           368       1738     1240          
17     2013  5     2  1926     1929            -3       2157     2310          
18     2013  5     4  1816     1820            -4       2017     2131          
19     2013  5     7  2054     2055            -1       2317       28          
20     2013  7    13  1947     1829            78       2143     2135          
21     2013  7    15  1921     1710           131       2106     2005          
22     2013  1     3  1957     1945            12       2232     2329          
23     2013  2    24   936      900            36       1507     1540          
24     2013  2    28   924      900            24       1455     1540          
25     2013  5     6  1036      730           186       1302     1105          
26     2013  5     6  1753     1755            -2       2004     2115          
27     2013  5    12  1744     1729            15       1948     2042          
28     2013  5    13  2044     2030            14       2301     2356          
29     2013  3     1  1929     1645           164       2141     2005          
30     2013  5     4  1914     1915            -1       2107     2216          
⋮      ⋮    ⋮     ⋮   ⋮        ⋮              ⋮         ⋮        ⋮             
336747 2013 9     25    NA     1655            NA         NA     1840          
336748 2013 9     25    NA     1559            NA         NA     1719          
336749 2013 9     26  1331     1329             2       1923     1813          
336750 2013 9     26    NA      915            NA         NA     1141          
336751 2013 9     26    NA     1400            NA         NA     1512          
336752 2013 9     26    NA     1240            NA         NA     1525          
336753 2013 9     27  1332     1329             3       1629     1509          
336754 2013 9     27  2253     1945           188         NA     2146          
336755 2013 9     27    NA      600            NA         NA      730          
336756 2013 9     27    NA     2100            NA         NA     2211          
336757 2013 9     27    NA     1329            NA         NA     1444          
336758 2013 9     27    NA     1600            NA         NA     1739          
336759 2013 9     28   555      600            -5        953      753          
336760 2013 9     28   847      839             8       1130      959          
336761 2013 9     28  1010     1020           -10       1344     1222          
336762 2013 9     28  1214     1225           -11       1801     1510          
336763 2013 9     28    NA     1803            NA         NA     1927          
336764 2013 9     28    NA      910            NA         NA     1220          
336765 

In [42]:
new_df<-mutate(flights,gain=dep_delay-arr_delay)

## Filtering extreme values
In the previous question we needed to find rows that had a large value of a certain column (`gain`). This occurs frequently, so the designers of tidyverse wrote a special function:

    top_n(<DATA FRAME>, n, <COLUMN>, ...)
    

In [44]:
# use top_n to find the flights with the highest gain
top_n(new_df,1,gain) # NOTE: 1 is the top 1 gain... you can change it to top 10 instead if you wanted to

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1 2013 6     13  1907     1512           235       2134     1928          
  arr_delay carrier flight tailnum origin dest air_time distance hour minute
1 126       EV      4377   N19554  EWR    JAX  126      820      15   12    
  time_hour           gain
1 2013-06-13 15:00:00 109

## Grouping data
Very frequently our data have natural groupings. For example, in flights, we might be interested in studying differences in flights depending on the month of departure. We use the `group_by()` function to tell R how to group data.

For example, `mtcars` is a dataset of cars and the gas mileage they get:

In [60]:
head(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

In [61]:
head(mtcars)%>%print # NOTE: if you don't specify no of records, first 6 records 

                   mpg cyl disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1


In [62]:
head(mtcars,5)%>%print # NOTE: first 5

                   mpg cyl disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2


In [63]:
?mtcars # NOTE: helps you understand what dataset is about

Let's try grouping `mtcars` by `cyl` (the number of engine cylinders):

In [64]:
print(group_by(mtcars, cyl)) # NOTE: nothing really happened, even when we did group_by

# A tibble: 32 × 11
# Groups:   cyl [3]
     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
 1  21       6  160    110  3.9   2.62  16.5     0     1     4     4
 2  21       6  160    110  3.9   2.88  17.0     0     1     4     4
 3  22.8     4  108     93  3.85  2.32  18.6     1     1     4     1
 4  21.4     6  258    110  3.08  3.22  19.4     1     0     3     1
 5  18.7     8  360    175  3.15  3.44  17.0     0     0     3     2
 6  18.1     6  225    105  2.76  3.46  20.2     1     0     3     1
 7  14.3     8  360    245  3.21  3.57  15.8     0     0     3     4
 8  24.4     4  147.    62  3.69  3.19  20       1     0     4     2
 9  22.8     4  141.    95  3.92  3.15  22.9     1     0     4     2
10  19.2     6  168.   123  3.92  3.44  18.3     1     0     4     4
# … with 22 more rows


In [65]:
gp<-group_by(mtcars, cyl)

This has not changed the data in any way. But now watch what happens when we use `mutate()` on the grouped data frame:

In [66]:
# mean mpg for grouped data
mutate(gp,mean(mpg))

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb mean(mpg)
1  21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4    19.74286 
2  21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4    19.74286 
3  22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1    26.66364 
4  21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1    19.74286 
5  18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2    15.10000 
6  18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1    19.74286 
7  14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4    15.10000 
8  24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2    26.66364 
9  22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2    26.66364 
10 19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4    19.74286 
11 17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4    19.74286 
12 16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3    15.10000 
13 17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3    15.10000 
14 15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3    15.10000 
15 10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4    15.10000 
16 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4    15.10000 
17 14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4    15.10000 
18 32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1    26.66364 
19 30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2    26.66364 
20 33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1    26.66364 
21 21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1    26.66364 
22 15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2    15.10000 
23 15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2    15.10000 
24 13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4    15.10000 
25 19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2    15.10000 
26 27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1    26.66364 
27 26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2    26.66364 
28 30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2    26.66364 
29 15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4    15.10000 
30 19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6    19.74286 
31 15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8    15.10000 
32 21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2    26.66364

In [86]:
distinct(mtcars, cyl) 

cyl
Mazda RX4         6  
Datsun 710        4  
Hornet Sportabout 8

Notice that the mean is now constant within different groups. It's easier to see if we first sort the table by `cyl`:

In [ ]:
# sort mtcars by cyl, then group and mutate
# FIND OUT ONCE POSTED AFTER CLASS 

## Pipes
Starting now, we will make extensive use of the pipe operator `%>%`. Consider the previous exercise:

In [71]:
# sort mtcars by cyl, then group, then mutate
arrange(
  select(
    mutate(flights,gain=dep_delay-arr_delay),
    gain, dep_delay, arr_delay,everything()),
  desc(gain)) # SELECT MOVES THINGS AROUND 

gain dep_delay arr_delay year month day dep_time sched_dep_time arr_time
1      109  235       126       2013  6    13  1907     1512           2134    
2       87   60       -27       2013  2    26  1000      900           1513    
3       80  206       126       2013  2    23  1226      900           1746    
4       79   17       -62       2013  5    13  1917     1900           2149    
5       76   24       -52       2013  2    27   924      900           1448    
6       74   48       -26       2013  7    14  1917     1829           2109    
7       74   34       -40       2013  7    17  2004     1930           2224    
8       73   31       -42       2013 12    27  1719     1648           1956    
9       73   -2       -75       2013  5     2  1947     1949           2209    
10      72    9       -63       2013 11    13  2024     2015           2251    
11      72  -14       -86       2013  5     7  1715     1729           1944    
12      72   19       -53       2013  5    12  2019     2000           2241    
13      71   39       -32       2013  5     2  1949     1910           2208    
14      71   -4       -75       2013  5     6  1826     1830           2045    
15      70    0       -70       2013  2    26  1335     1335           1819    
16      70  368       298       2013  3     1  1528      920           1738    
17      70   -3       -73       2013  5     2  1926     1929           2157    
18      70   -4       -74       2013  5     4  1816     1820           2017    
19      70   -1       -71       2013  5     7  2054     2055           2317    
20      70   78         8       2013  7    13  1947     1829           2143    
21      70  131        61       2013  7    15  1921     1710           2106    
22      69   12       -57       2013  1     3  1957     1945           2232    
23      69   36       -33       2013  2    24   936      900           1507    
24      69   24       -45       2013  2    28   924      900           1455    
25      69  186       117       2013  5     6  1036      730           1302    
26      69   -2       -71       2013  5     6  1753     1755           2004    
27      69   15       -54       2013  5    12  1744     1729           1948    
28      69   14       -55       2013  5    13  2044     2030           2301    
29      68  164        96       2013  3     1  1929     1645           2141    
30      68   -1       -69       2013  5     4  1914     1915           2107    
⋮      ⋮    ⋮         ⋮         ⋮    ⋮     ⋮   ⋮        ⋮              ⋮       
336747 NA    NA       NA        2013 9     25    NA     1655             NA    
336748 NA    NA       NA        2013 9     25    NA     1559             NA    
336749 NA     2       NA        2013 9     26  1331     1329           1923    
336750 NA    NA       NA        2013 9     26    NA      915             NA    
336751 NA    NA       NA        2013 9     26    NA     1400             NA    
336752 NA    NA       NA        2013 9     26    NA     1240             NA    
336753 NA     3       NA        2013 9     27  1332     1329           1629    
336754 NA   188       NA        2013 9     27  2253     1945             NA    
336755 NA    NA       NA        2013 9     27    NA      600             NA    
336756 NA    NA       NA        2013 9     27    NA     2100             NA    
336757 NA    NA       NA        2013 9     27    NA     1329             NA    
336758 NA    NA       NA        2013 9     27    NA     1600             NA    
336759 NA    -5       NA        2013 9     28   555      600            953    
336760 NA     8       NA        2013 9     28   847      839           1130    
336761 NA   -10       NA        2013 9     28  1010     1020           1344    
336762 NA   -11       NA        2013 9     28  1214     1225           1801    
336763 NA    NA       NA        2013 9     28    NA     1803             NA    
336764 NA    NA       NA        2013 9     28    NA      910             NA    
336765 

This is not very nice. To figure out what the command is doing you have to work from the inside out, which is not the order in which we are accustomed to reading. A slight improvement might be:

In [ ]:
# sort mtcars by cyl, then group, then mutate, using intermediates

This is better, but now you've created a bunch of useless temporary variables, and it requires a lot of typing. 
Instead, we are going to use a new operator `%>%` (prounouced "pipe"):

In [ ]:
# sort mtcars by cyl, then group, then mutate, using pipes
mutate(flights,gain=dep_delay-arr_delay) %>% 
  select(gain,dep_delay,arr_delay,everything()) %>% 
    arrange(desc(gain)) 
    # NOTE: PIPING IS SO COOL! DON'T NEED TO WORRY ABOUT NESTED FUNCTIONS

In [78]:
new_df1=mutate(flights,gain=dep_delay-arr_delay)
new_df2=select(new_df1,gain,dep_delay,arr_delay,everything())
new_df3=arrange(new_df2,desc(gain))
print(new_df2)
# NOTE: these r the step by step. makes more logical sense, is more intuitive. 
# but biggest disadvantage of doing this is that you're wasting memory.

# A tibble: 336,776 × 20
    gain dep_delay arr_delay  year month   day dep_time sched_…¹ arr_t…² sched…³
   <dbl>     <dbl>     <dbl> <int> <int> <int>    <int>    <int>   <int>   <int>
 1    -9         2        11  2013     1     1      517      515     830     819
 2   -16         4        20  2013     1     1      533      529     850     830
 3   -31         2        33  2013     1     1      542      540     923     850
 4    17        -1       -18  2013     1     1      544      545    1004    1022
 5    19        -6       -25  2013     1     1      554      600     812     837
 6   -16        -4        12  2013     1     1      554      558     740     728
 7   -24        -5        19  2013     1     1      555      600     913     854
 8    11        -3       -14  2013     1     1      557      600     709     723
 9     5        -3        -8  2013     1     1      557      600     838     846
10   -10        -2         8  2013     1     1      558      600     753     745
# …

This is much better. We can read the command from left to right and know exactly what is going on.

## Column operation #3: `summarize()`ing data

`summarize()` can be used to summarize entire data frames by collapsing them into single number summaries. The syntax is:

    summarize(<grouped data frame>, 
              <new variable> = <formula for new variable>,
              <other new variable> = <other formula>)

The most basic use of summarize is to compute statistics over the whole data set:

In [81]:
# summarize flights by mean of departure delay
summarize(flights,mean(dep_delay))
# anything that NA touches becomes NA 

mean(dep_delay)
1 NA

In [83]:
# you could give the column a name
summarize(flights,mean_dep_delay=mean(dep_delay))

mean_dep_delay
1 NA

In [82]:
summarize(flights,mean(dep_delay,na.rm=T)) #NA REMOVE IS TRUE

mean(dep_delay, na.rm = T)
1 12.63907

`summarize()` applies a summary function to each group of data. Remember that it always returns **one row per group**. In the above example, there was only one group (the whole data set), so the resulting data frame had only one row.

### Grouping observations
`summarize()` is most useful when combined with `group_by()` to group observations before calculating the summary statistic. Let's summarize flights by the mean departure delay in each month.

In [85]:
# summarize average departure delay by month.
group_by(mtcars,cyl)%>%
  summarize(mean_mpg=mean(mpg))%>% 
  print

# A tibble: 3 × 2
    cyl mean_mpg
  <dbl>    <dbl>
1     4     26.7
2     6     19.7
3     8     15.1


### Example: counting the number of rows
The `n()` function calculates the number of rows in each group:

In [ ]:
# count the number of rows in flights for each month

### A shortcut
`summarize(n = n())` occurs so often that there is a shortcut for it:

In [87]:
# use count() instead of group_by or summarize
count(mtcars,cyl) 
# NOTE: count w/o the cyl gives u total no of obs. but now, gives u the count per cyl value. the distinct values & the total count for each value.

cyl n 
1 4   11
2 6    7
3 8   14

Let's think about how to answer the following question using `summarize`:

## What days of the year / at what airport are the busiest for flying?

To figure this out, I like to think about/visualize the table we would want to have in order to easily answer this question. Ideally, it would look something like this:

    # A tibble: 1,095 x 4
       month   day airport       n_departures
       <int> <int> <chr>                <int>
     1     1     1 EWR                    305
     2     1     1 JFK                    297
     3     1     1 LGA                    240
     4     1     2 EWR                    350
     5     1     2 JFK                    321
     6     1     2 LGA                    272
     7     1     3 EWR                    336
     8     1     3 JFK                    318
     9     1     3 LGA                    260
    10     1     4 EWR                    339
    # … with 1,085 more rows

Then, to get the answer, I could sort the table to find the row that had the largest `n_departures`.

How do I reach the table shown above? There is one row per ... what? (This tells me how to group the data.)

In [90]:
# summarize flights to get number of departures by day and by airport.
group_by(flights,month,day,origin) %>% 
  summarize(n=n()) %>% 
    print

`summarise()` has grouped output by 'month', 'day'. You can override using the
`.groups` argument.


# A tibble: 1,095 × 4
# Groups:   month, day [365]
   month   day origin     n
   <int> <int> <chr>  <int>
 1     1     1 EWR      305
 2     1     1 JFK      297
 3     1     1 LGA      240
 4     1     2 EWR      350
 5     1     2 JFK      321
 6     1     2 LGA      272
 7     1     3 EWR      336
 8     1     3 JFK      318
 9     1     3 LGA      260
10     1     4 EWR      339
# … with 1,085 more rows


Here is another question we can answer:

## Who is the greatest (baseball) batter of all time?
The `Lahman` dataset contains information on baseball players.

In [91]:
install.packages("Lahman")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [92]:
# install.packages("Lahman")
library(Lahman)
as_tibble(Batting) %>% head
# what do all these columns mean?

playerID  yearID stint teamID lgID G  AB  R  H  X2B ⋯ RBI SB CS BB SO IBB HBP
1 abercda01 1871   1     TRO    NA    1   4  0  0  0  ⋯  0  0  0  0  0  NA  NA 
2 addybo01  1871   1     RC1    NA   25 118 30 32  6  ⋯ 13  8  1  4  0  NA  NA 
3 allisar01 1871   1     CL1    NA   29 137 28 40  4  ⋯ 19  3  1  2  5  NA  NA 
4 allisdo01 1871   1     WS3    NA   27 133 28 44 10  ⋯ 27  1  1  0  2  NA  NA 
5 ansonca01 1871   1     RC1    NA   25 120 29 39 11  ⋯ 16  6  2  2  1  NA  NA 
6 armstbo01 1871   1     FW1    NA   12  49  9 11  2  ⋯  5  0  1  0  1  NA  NA 
  SH SF GIDP
1 NA NA 0   
2 NA NA 0   
3 NA NA 1   
4 NA NA 0   
5 NA NA 0   
6 NA NA 0

The second player is `addybo01`. We can get information about this player by typing:

In [94]:
Lahman::playerInfo('addybo01')

playerID nameFirst nameLast
111 addybo01 Bob       Addy

Bob Addy was active in the years 1871-1877. During that time he had $118+51+152+213+310+142+245=1231$ at-bats, and $32+16+54+51+80+40+68=341$ hits. Therefore his career batting average was $341/1241=0.277$.

In [95]:
filter(Batting, playerID == 'addybo01')

playerID yearID stint teamID lgID G  AB  R  H  X2B ⋯ RBI SB CS BB SO IBB HBP
1 addybo01 1871   1     RC1    NA   25 118 30 32 6   ⋯ 13   8  1 4  0  NA  NA 
2 addybo01 1873   1     PH2    NA   10  51 12 16 1   ⋯ 10   1  1 2  0  NA  NA 
3 addybo01 1873   2     BS1    NA   31 152 37 54 6   ⋯ 32   6  5 2  1  NA  NA 
4 addybo01 1874   1     HR1    NA   50 213 25 51 9   ⋯ 22   4  2 1  1  NA  NA 
5 addybo01 1875   1     PH2    NA   69 310 60 80 8   ⋯ 43  16  8 0  2  NA  NA 
6 addybo01 1876   1     CHN    NL   32 142 36 40 4   ⋯ 16  NA NA 5  0  NA  NA 
7 addybo01 1877   1     CN1    NL   57 245 27 68 2   ⋯ 31  NA NA 6  5  NA  NA 
  SH SF GIDP
1 NA NA  0  
2 NA NA  0  
3 NA NA  0  
4 NA NA  0  
5 NA NA  0  
6 NA NA NA  
7 NA NA NA

Let's use `group_by()` and `summarize()` to calculate the "career" batting average for every player in the dataset. That is, I want a table that looks like:

    # A tibble: 20166 × 2
      playerID batting_avg
      <chr>          <dbl>
    1 addybo01       0.277
    .    .             .
    .    .             .
    .    .             .

In [103]:
# calculate the batting average for each player in the  data set
group_by(Batting,playerID)%>%
  summarize('batting avg'=sum(H)/sum(AB))%>%
  top_n(10)

Selecting by batting avg


playerID  batting avg
1  abramge01 1          
2  alberan01 1          
3  banisje01 1          
4  bartocl01 1          
5  bassdo01  1          
6  birasst01 1          
7  bruneju01 1          
8  burnscb01 1          
9  cammaer01 1          
10 campsh01  1          
11 crockcl01 1          
12 davenjo01 1          
13 davidda01 1          
14 devinha01 1          
15 dupremi01 1          
16 eddydo01  1          
17 egeco01   1          
18 eubanue01 1          
19 fryjo01   1          
20 gallaja01 1          
21 garcimi02 1          
22 garcira01 1          
23 gideobr01 1          
24 gleasro01 1          
25 gowella01 1          
26 grasmlo01 1          
27 grayda01  1          
28 hancory01 1          
29 hesslke01 1          
30 holloji01 1          
⋮  ⋮         ⋮          
66 pickuty01 1          
67 pindebr01 1          
68 rambope01 1          
69 roachja01 1          
70 rodried01 1          
71 rowlami01 1          
72 schemfr01 1          
73 schlibr01 1          
74 schlogu01 1          
75 sellsda01 1          
76 shirlst01 1          
77 simspe01  1          
78 skidmro01 1          
79 smithwi04 1          
80 snowch01  1          
81 sotogr01  1          
82 stoneti01 1          
83 suchech01 1          
84 swartbu01 1          
85 tompkch01 1          
86 tonkido01 1          
87 toppiru01 1          
88 torrejo02 1          
89 tupmama01 1          
90 wattal01  1          
91 wetzebu01 1          
92 whiteab01 1          
93 wrighal01 1          
94 yanes01   1          
95 yantzge01 1

What has happened? Let's look at the first player in the sorted table:

In [104]:
filter(Batting, playerID == 'abramge01')

playerID  yearID stint teamID lgID G AB R H X2B ⋯ RBI SB CS BB SO IBB HBP SH
1 abramge01 1923   1     CIN    NL   3 1  0 1 0   ⋯ 0   0  0  0  0  NA  0   0 
  SF GIDP
1 NA NA

### Always include counts
It is a good idea to include counts of each group when you do a summary. Some groups may have very low numbers of observations, resulting in high variance for the summary statistics. 

What happens if we restrict our batting average calculation to players that had at least 100 at-bats?

## Quiz
Among players who had at least 100 at bats, who had the highest career batting average?
<ol style="list-style-type: upper-alpha;">
    <li>Ty Cobb</li>
    <li>Babe Ruth</li>
    <li>Prof. Terhorst</li>
    <li>Ted Williams</li>
    <li>Hank Williams</li>
</ol>

In [ ]:
# highest batting average among players that had 100 or more at bats

## Quiz
Among players who had at least 100 at bats in a season, what was the highest batting average in a single season?
<ol style="list-style-type: upper-alpha;">
    <li>Ted Williams</li>
    <li>Steven Colbert</li>
    <li>Chonky Squirrel</li>
    <li>Levi Meyerle</li>
    <li>Tom Riddle</li>
</ol>

In [ ]:
# highest seasonal batting average

## The Steroid Era of Baseball

> [Baseball] remained relatively the same until the 90s when steroid use became rampant. Famous sluggers like Barry Bonds, Mark McGwire, and Sammy Sosa rose to fame during this era. They were beloved at the time until we later found out that they were cheating.

https://www.wagerbop.com/how-home-runs-and-batting-averages-have-changed-over-the-last-30-years/

![barry bounds](https://cdn.ebaumsworld.com/mediaFiles/picture/2605038/87087115.jpg)

## Can we see the steroid era reflected in the data?

In [ ]:
# summarize the dataset in order to investigate steroid era in batters